src/Core/ScreenshotCache.vala

```diff
@@ -0,0 +1,201 @@
+                    var size = (int) info.get_size ();
+                    file.delete ();
+                    AtomicInt.set (ref screenshot_usage, AtomicInt.get (ref screenshot_usage) - size);
+                } catch (Error e) {
+                    stderr.printf ("failed to delete %s: %s\n", oldest_path, e.message);
+                }
+            }
+        }
+    }
+
+    // Get the combined size of the screenshot cache.
+    private int summarize_screenshot_usage () {
+        Dir dir = Dir.open (screenshot_path, 0);
+        string? name = null;
+        int size = 0;
+
+        while ((name = dir.read_name ()) != null) {
+            string entry = Path.build_filename (screenshot_path, name);
+            File file = File.new_for_path (entry);
+            FileInfo info = file.query_info ("standard::*", FileQueryInfoFlags.NONE);
+            size += (int) info.get_size ();
+        }
+
+        return size;
+    }
+
+    /*
+     * Fetches a screenshot in a background thread.
+     * 
+     * A result indicating the success (0) will be returned as the result upon completion.
+     */
+    public async int fetch (string url, out File out_file) {
+        SourceFunc callback = fetch.callback;
+        int ext_pos = url.last_index_of (".");
+        string extension = url.slice ((long) ext_pos, (long) url.length);
+        string file_name = "%02x".printf (url.hash ());
+
+        string path = screenshot_path + "/" + file_name + extension;
+        int result = 0;
+
+        var file = File.new_for_path (path);
+
+        new Thread<bool> ("fetching_screenshot", () => {
+            FileIOStream stream;
+            var session = new Soup.Session ();
+            bool download = true;
+            time_t mtime = 0;
+
+            try {
+                if (file.query_exists ()) {
+                    stream = file.open_readwrite ();
+                    var msg = new Soup.Message ("HEAD", url);
+                    session.send_message (msg);
+
+                    // Compare the mtimes of the header and the existing file.
+                    // If they're the same, we do not need to download it again.
+                    var modified = msg.response_headers.get_one ("Last-Modified");
+                    if (null != modified) {
+                        var time = new Soup.Date.from_string (modified).to_time_t ();
+                        if (Stat (path).st_mtime != time) {
+                            mtime = time;
+                        } else {
+                            download = false;
+                        }
+                    }
+                } else {
+                    stream = file.create_readwrite (FileCreateFlags.NONE);
+                }
+            } catch (Error e) {
+                debug (e.message);
+                result = -1;
+                return true;
+            }
+
+            if (download) {
+                stderr.printf ("downloading %s to %s\n", url, path);
```

Comment:
Should probably be a `debug ()` call or just remove it entirely.

Reply:
I haven't figured out how to get `debug ()` to print anything. Is there something special that needs to be done when compiling it?

Reply:
@mmstick when running, you can set the environment variable `G_MESSAGE_DEBUG` to the debug domain, or just set it to `all`. So `G_MESSAGE_DEBUG=all io.elementary.appcenter` in this case.

Reply:
@cassidyjames For whatever reason, neither `env G_MESSAGE_DEBUG=debug ./src/io.elementary.appcenter` nor `env G_MESSAGE_DEBUG=all ./src/io.elementary.appcenter` seem to work.

Reply:
@mmstick my bad, I had a typo. It's `G_MESSAGES_DEBUG`.